pkg/tcpip/network/ipv4/icmp.go

```diff
@@ -134,6 +134,44 @@ func (e *endpoint) handleICMP(r *stack.Route, netHeader buffer.View, vv buffer.V
 	case header.ICMPv4Timestamp:
 		received.Timestamp.Increment()
 
+		// Only send a reply if the checksum is valid.
+		wantChecksum := h.Checksum()
+		// Reset the checksum field to 0 to can calculate the proper
+		// checksum. We'll have to reset this before we hand the packet off.
+		h.SetChecksum(0)
+
+		// It's possible that a raw socket expects to receive this.
+		h.SetChecksum(wantChecksum)
+		e.dispatcher.DeliverTransportPacket(r, header.ICMPv4ProtocolNumber, netHeader, vv)
+
+		gotChecksum := ^header.ChecksumVV(vv, 0 /* initial */)
+		if gotChecksum != wantChecksum {
+			received.Invalid.Increment()
+			return
+		}
+
+		vv := vv.Clone(nil)
+		vv.TrimFront(header.ICMPv4TimeStampMinimumSize)
+		hdr := buffer.NewPrependable(int(r.MaxHeaderLength()) + header.ICMPv4TimeStampMinimumSize)
+		pkt := header.ICMPv4(hdr.Prepend(header.ICMPv4TimeStampMinimumSize))
+		copy(pkt, h)
+		pkt.SetType(header.ICMPv4TimestampReply)
+
+		// millisecondsPerDay is Milli Seconds in a day calculation for Timestamp reply messages.
+		lastDay := now.Truncate(24 * time.Hour)
+		ms := now.Since(lastDay).Milliseconds()
+		timeStamp := ((time.Now().UnixNano() / 1000000) % ms)
+		pkt.SetRxTimeStamp(timeStamp, timeStamp)
+
```

Comment:
We seem to be missing setting the txTimeStamp in the reply packet. As per RFC 777

 0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |     Type      |                  unused                       |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |     Originate Timestamp                                       |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |     Receive Timestamp                                         |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |     Transmit Timestamp                                        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

   IP Fields:

   Addresses

      The address of the source in a timestamp message will be the
      destination of the timestamp reply message.  To form a timestamp
      reply message, the source and destination addresses are simply
      reversed.

   IP Fields:

   Type

      13 for timestamp message;

      14 for timestamp reply message.

   Description

      The data received (a timestamp) in the message is returned in the
      reply together with an additional timestamp.  The timestamp is 32
      bits of milliseconds since midnight UT.  One use of these
      timestamps is described by Mills [5].